src/org/omegat/gui/editor/EditorController.java

```diff
@@ -569,54 +570,29 @@ private void setFont(final Font font) {
      * Decide what document orientation should be default for source/target languages.
      */
     private void setInitialOrientation() {
-        String sourceLang = Core.getProject().getProjectProperties().getSourceLanguage().getLanguageCode();
-        String targetLang = Core.getProject().getProjectProperties().getTargetLanguage().getLanguageCode();
-
-        sourceLangIsRTL = Language.isRTL(sourceLang);
-        targetLangIsRTL = Language.isRTL(targetLang);
+        getValuesToDetermineOrientation();
+        applyOrientationToEditor(currentOrientation);
+    }
 
-        if (sourceLangIsRTL != targetLangIsRTL || sourceLangIsRTL != Language.localeIsRTL()) {
-            currentOrientation = Document3.ORIENTATION.DIFFER;
-        } else {
-            if (sourceLangIsRTL) {
-                currentOrientation = Document3.ORIENTATION.ALL_RTL;
-            } else {
-                currentOrientation = Document3.ORIENTATION.ALL_LTR;
-            }
-        }
-        applyOrientationToEditor();
+    private void getValuesToDetermineOrientation() {
+        sourceLangIsRTL = BiDiUtils.isSourceLangRtl();
+        targetLangIsRTL = BiDiUtils.isTargetLangRtl();
+        currentOrientation = BiDiUtils.getOrientationType();
     }
 
     /**
      * Define editor's orientation by target language orientation.
      */
-    private void applyOrientationToEditor() {
-        ComponentOrientation targetOrientation = null;
-        switch (currentOrientation) {
-        case ALL_LTR:
-            targetOrientation = ComponentOrientation.LEFT_TO_RIGHT;
-            break;
-        case ALL_RTL:
-            targetOrientation = ComponentOrientation.RIGHT_TO_LEFT;
-            break;
-        case DIFFER:
-            if (targetLangIsRTL) { //using target lang direction gives better result when user starts editing.
-                targetOrientation = ComponentOrientation.RIGHT_TO_LEFT;
-            } else {
-                targetOrientation = ComponentOrientation.LEFT_TO_RIGHT;
-            }
-        }
-        // set editor's orientation by target language
-        editor.setComponentOrientation(targetOrientation);
+    private void applyOrientationToEditor(BiDiUtils.ORIENTATION orientation) {
+        editor.setComponentOrientation(BiDiUtils.getOrientation(orientation));
     }
 
     /**
-     * returns the orientation of the document
-     * (so we can decide what way of tag colouring we need;
-     * if that has been fixed in an other way, this method can be removed again.).
-     * @return
+     * returns the orientation of the document (so we can decide what way of tag
+     * colouring we need; if that has been fixed in an other way, this method
+     * can be removed again.).
      */
-    public Document3.ORIENTATION getOrientation() {
+    public BiDiUtils.ORIENTATION getOrientation() {
```

Comment:
Here is a place where public method value type changed. It can be compatibility breaking change.
If there is strong reason to break compatibility, we can keep  it to return `Document3.ORIENTATION`

Reply:
I think we can define

```
    /**
     * The orientation of the document is all RTL.
     * @return true when the orientation is all RtL. otherwise false.
     */
    public boolean isOrientationAllRtl() {
        return currentOrientation.equals(BiDiUtils.ORIENTATION.ALL_RTL);
    }
```

We can avoid return ORIENTATION enum from method.